### ISSUE TITLE
Recursion Error when Differentiating Functions

### ISSUE THREAD

**Issue by BobatPurdue** [sympy/sympy#13791]
I'm using sympy 1.1.1 on Anaconda 4.4.0.  I've tried differentiating a function of functions.  It will crash python.  In other circumstances, I will get a recursion error.  Anyone else see this problem?  Suggestions?

```
from sympy import Symbol, Function

x = Symbol('x')
y = Symbol('y')
u = Function('u')(x,y)
v = Function('v')(x,y,u)

Okay = v.diff(y).diff(x)
print(Okay)
Problem = Okay.diff(y)
print(Problem)
```

**Comment by BobatPurdue** [comment]
Just tried 1.1.2.dev.  I had the same problem.

**Comment by normalhuman** [comment]
I think "Okay" might not be okay, either... First, if you change the order of differentiation producing Okay (symbolic derivatives don't depend on the order), then the following works without errors:
```
Okay = v.diff(x).diff(y)    # instead of  y, x
print(Okay)
Problem = Okay.diff(y)
print(Problem)              # no problem (?)
```
Disturbingly, v.diff(x).diff(y) and v.diff(y).diff(x) are different; and at this moment I'm not sure that either one is correct.

**Comment by normalhuman** [comment]
Here is a simplified example, which also shows where the infinite recursion comes from. 
```
x = Symbol('x')
u = Function('u')
Derivative(u(x, u(x, x)), x, x).doit()
```
The answer is a long sum, but the last term in that sum is `Derivative(u(x, u(x, x)), x, x)` which is precisely the thing we started with. Hence, the answer is incorrect and is potential for infinite recursion. 
That last term should have been `Subs(Derivative(u(x, _xi), x, x), _xi, u(x, x))` with a dummy symbol replacing u(x, x) to hold that term constant during differentiation. 

So it seems that second and higher derivatives of compositions of unknown functions cannot be trusted at present.

**Comment by normalhuman** [comment]
I found the step at which this error is made, and isolated it into a simpler example:  
```
x, y = symbols('x y')
u = Function('u')
Subs(Derivative(u(y, u(x, 0)), y), y, x).diff(x)
```
The output from current master is 
```
Derivative(u(x, 0), x)*Derivative(u(x, u(x, 0)), u(x, 0), x) + Derivative(u(x, u(x, 0)), (x, 2))
```
The first term is correct, but the second is not: it should be `Subs(Derivative(u(x, _xi), (x, 2)), _xi, u(x, 0))`. 

Using the subscript notation for partial derivatives: the expression we started with was u_1 (x, u(x, 0)). Its derivative wrt x is `u_1(x, 0) * u_{12}(x, u(x, 0)) + u_{11}(x, u(x, 0))`.  It's the last term `u_{11}(x, u(x, 0))` that isn't expressed correctly.
